jupyter-workpad/thomas_scratchpad/Unicom-training/retrieval.py

```diff
@@ -0,0 +1,1056 @@
+#     )
+
+
+#     module_partial_fc = PartialFC_V2(
+#         margin_loss, 
+#         args.output_dim,
+#         dataset_train.num_classes,
+#         args.sample_rate,
+#         False,
+#         sample_num_feat=args.num_feat
+#     )
+#     module_partial_fc.train().cuda()
+
+#     # Initialize optimizer
+#     opt = torch.optim.AdamW(
+#         params=[
+#             {"params": backbone.parameters()},
+#             {"params": module_partial_fc.parameters(), "lr": args.lr * args.lr_pfc_weight}
+#         ],
+#         lr=args.lr,
+#         weight_decay=args.weight_decay
+#     )
+
+#     # Learning rate scheduler
+#     steps_per_epoch = len(dataset_train) // (world_size * args.batch_size) + 1
+#     lr_scheduler = optim.lr_scheduler.OneCycleLR(
+#         optimizer=opt,
+#         max_lr=[args.lr, args.lr * args.lr_pfc_weight],
+#         steps_per_epoch=steps_per_epoch,
+#         epochs=args.epochs,
+#         pct_start=0.1,
+#     )
+
+#     # Training loop metrics
+#     callback_func = SpeedCallBack(10, args.epochs * steps_per_epoch, args.batch_size)
+#     auto_scaler = torch.cuda.amp.grad_scaler.GradScaler(growth_interval=200)
+#     global_step = 0
+
+#     # Main training loop
+#     for epoch in range(args.epochs):
+#         if train_sampler is not None:
+#             train_sampler.set_epoch(epoch)
+            
+#         for _, (img, local_labels) in enumerate(loader_train):
+#             img = img.cuda()
+#             local_labels = local_labels.long().cuda()
+            
+#             with torch.cuda.amp.autocast(False):
+#                 local_embeddings = backbone(img)
+#             local_embeddings.float()
+            
+#             loss = module_partial_fc(local_embeddings, local_labels)
+#             auto_scaler.scale(loss).backward()
+
+#             if global_step % args.gradient_acc == 0:
+#                 auto_scaler.step(opt)
+#                 auto_scaler.update()
+#                 opt.zero_grad()
+
+#             lr_scheduler.step()
+#             global_step += 1
+
+#             with torch.no_grad():
+#                 callback_func(
+#                     lr_scheduler,
+#                     float(loss),
+#                     global_step,
+#                     auto_scaler.get_scale()
+#                 )
+
+#         if rank == 0:
+#             print(f"Completed epoch {epoch}")
+
+#     if rank == 0:
+#         print("Training completed")
+
+
+def main(config=None):
+    print("Current working directory:", os.getcwd())
+    if config is None:
+        config = Config()
+    
+    # Make config globally available as args
+    global args
+    args = config
+    
+    # Initialize distributed processing with a single process
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+    distributed.init_process_group(backend='nccl', world_size=1, rank=0)
+
+    
+    # Load model and initialize
+    model, transform_clip = clip.load(args.model_name)
```

Comment:
Do I need to use the original unicom-provided models?  This line requires some additional investigation from me.

Reply:
Current conclusion is to use a pretrained base model, e.g., `ViT/B-32`.